### PR TITLE
application: CILA - omnichain ifrastructure

### PR DESCRIPTION
**Pull request by AlexShkor**
#### Project Abstract

The goal of this project is to implement an omnichain smart contract infrastructure support for Substrate framework. This will include the implementation of CQRS + Event Sourcing execution environment plus an example smart-contract. The implementation will be done using WASM and/or native Rust Substrate implementation as a pallet. The implementation will include implementing Protobuf support on-chain, serialization/deserialization, aggregated repository, event store, command/operations dispatcher, and events emitter.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

File: applications/CILA-omnichain-infrastructure.md [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+
+## Usage
+
+The `Dispatcher` can be used in the implementation of the CQRS + Event Sourcing execution environment for Substrate framework. It receives commands/operations from external sources, such as a client or a node, and routes them to the appropriate command/operation handler.
+
+To use the `Dispatcher`, developers first define a set of command/operation handlers for the smart contract. These handlers can be implemented as methods in a Rust struct. The `Dispatcher` class then instantiates this struct and routes commands/operations to the appropriate method based on the type of command/operation.
+
+## Benefits
+
+The `Dispatcher` provides a simple and flexible way to handle commands/operations in the smart contract. By defining a set of command/operation handlers, developers can easily add new functionality to the smart contract without having to modify the `Dispatcher` class` itself.
```
Comment (takahser) [inline applications/CILA-omnichain-infrastructure.md:79]:
```suggestion
The `Dispatcher` provides a simple and flexible way to handle commands/operations in the smart contract. By defining a set of command/operation handlers, developers can easily add new functionality to the smart contract without having to modify the `Dispatcher` class itself.
```

File: applications/CILA-omnichain-infrastructure.md [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+- `get_version()`: Returns the version of the aggregate object.
+- `get_events()`: Returns the list of events that have occurred in the past.
+
+### Usage
+
+To use the `AggregateState` , you must first create an instance of it and initialize it with the current state of the aggregate object. You can then apply events to the aggregate object by calling the `apply_event()` method.
+
+## Diagramm
+
+Architcture Overview Diagramm
```
Comment (takahser) [inline applications/CILA-omnichain-infrastructure.md:196]:
```suggestion
Architecture Overview Diagram
```

File: applications/CILA-omnichain-infrastructure.md [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+
+Flow Diagramm
+
+![Diagramm](https://i.imgur.com/1jeLNo1.png)
+
+Command Processing Flow Diagramm
+
+```sql
+ +------------+       +------------+       +-----------------+       +-------------------+       +------------+
+ | Application|       |   Router   |       |Execution Chain  |       |  Event Relay Node |       | Aggregation|
```
Comment (takahser) [inline applications/CILA-omnichain-infrastructure.md:207]:
```suggestion
 | Application|       |   Router   |       | Execution Chain |       |  Event Relay Node |       | Aggregation|
```

**Review by takahser — CHANGES_REQUESTED** [review]
@AlexShkor thanks for submitting your grant proposal.
I've got a couple of questions:

- regarding "events relay" for cross-chain communication, what tech are you planning to use here? Protobuf?
- If yes, is there a reason not to use XCM here?
- Are you aiming to replace [Substrates' storage APIs](https://docs.substrate.io/build/runtime-storage/) with mongodb for the scope of your project?
- If yes, what's the reasoning behind that?
- What's your strategy on syncing the mongodb state with other blockchain nodes?
- Storing all event will make it much easier to interact with these smart contracts. However, this comes at a cost: loads of additional storage. Did you think of ways to ensure scalability here, when compared to the more "traditional" model where events are not necessarily persisted after their emittance?
- Why did you add NFTs to the architecture diagram? Are they that elementary to the architecture? I would have imagined them to be just one of many things you can implement through the smart contract.
- Regarding M2, we don't usually support deployment efforts. Feel free to remove it or to add some code deliveries, such as automated tests, fuzz tests or other code-related deliverables.
- Regarding M3, are you trying to reuse any existing NFT standards here? If not, why?

**Comment by AlexShkor** [comment]
@takahser 

`I imagine it to be quite expensive. Is the event emittance being paid for somehow by the contract owner or the person who calls a corresponding function that emits the event?`

yes, it will be paid by the original sender, the events will be batched with other smart contracts so it might help optimize the cost.

`By that, do you mean that it's a problem that an NFT can be part of a collection? Afaik there are both type of standards: Those which are part of a collection and those which aren't. If you seek funding for the NFT development it would be good to learn what additional value your implementation brings to the ecosystem and why it's not viable to just reuse an existing standard/implementation.`

Most probably we will reuse existing standards, I just wanted to mention here our approach to choosing a standard that we aim to reimplement in our omnichain format.

**Comment by takahser** [comment]
Thanks for the updates, I'm generally open to support M1 and M2.

> Most probably we will reuse existing standards, I just wanted to mention here our approach to choosing a standard that we aim to reimplement in our omnichain format.

If you run your own standard, I'd expect you to justify why it's needed. On the other hand, if it's feasible to reuse an existing standard/implementation it would be the better option. However, in my personal opinion, it wouldn't really deliver any additional value to the ecosystem, and it's questionable if we should fund this. An alternative idea would be to deliver a market research report on existing NFT standards/implementations in the ecosystem as part of M3, which might help other teams in a similar situation to make a choice. In that case, I'd personally be open to support it.

**Comment by AlexShkor** [comment]
@takahser Thank you, it's actually a good idea since we will be doing it anyway to select a proper standard for the omnichain implementation.

**Review by takahser — APPROVED** [review]
@AlexShkor thanks for the updates, I'm happy to vote in favor of your application now. I'm going to share it with the other committee members, please note that it may take some time before you'll receive their feedback as we have currently a lot of open applications that need to be reviewed. Thank you for your patience in advance.

**Review by keeganquigley — APPROVED** [review]
Thanks for the deep dive @takahser and for making the changes @AlexShkor CQRS sounds like interesting tech, and a good use case for an alternative WASM implementation. LGTM.

**Review by Noc2 — APPROVED** [review]
You clearly put a lot of effort into this. I'm happy to support it.

**Pull request merged**